### DISCUSSION TITLE
Why is `pnpm dlx <package>@latest` not bypassing cache?

### DISCUSSION THREAD

**Discussion by devjiwonchoi** [pnpm/pnpm#8630] [Q&A]
From this PR https://github.com/pnpm/pnpm/pull/7835 `pnpm dlx` is enabled with cache for [a DAY](https://github.com/pnpm/pnpm/blob/884fdd7dec9a453230d876d8c766a693e1683e12/config/config/src/index.ts#L153), but when we run a package with `@latest`, we expect to fetch the latest package.

1. Is there a reason why we are using cache even with `@latest`?
2. How can we workaround this behavior? (within the script)

**Comment by AlphaFoxz** [comment]
Same problem.
And here is an example to show how counter-intuitive it is.

## Step 1: Execute an example package

```sh
$ pnpm info create-example@latest
# Got latest: 0.0.1
$ pnpm dlx create-example # or pnpm create example@latest
```
Now, we installed `create-example@0.0.1` into [cache-dir]

## Step 2: Update example package

Modify `create-example`'s version to `0.0.2`. Then publish.

## Step 3: Execute the example package with newer, exact version

```sh
$ pnpm dlx create-example@0.0.2 # or pnpm create example@0.0.2
```
This time v0.0.2 be installed into [cache-dir]

## Step 4: Try execute example@latest

```sh
$ pnpm dlx create-example@latest # or pnpm create example@latest
```

## Finally, the process `create-example@0.0.1` will be executed again. Even both v0.0.1 and v0.0.2 are all cached.

**Comment by zkochan** [comment]
> but when we run a package with @latest, we expect to fetch the latest package.

I have no idea what is expected. One of the most upvoted issues was to make dlx fast by using a cache.

Any addition rules will make it slower. You can always run it with an exact version. I doubt it is a frequent scenario.

This should currently work to skip cache:

```
pnpm --config.dlx-cache-max-age=0 dlx ...
```

cc @KSXGitHub

**Reply by KSXGitHub** [reply]
We can skip cache entirely should a single `@latest` is found in packages list. It does mean that `dlx pkg@latest` will always be slow, but the user should know what they are doing anyway.

**Reply by AlphaFoxz** [reply]
> We can skip cache entirely should a single `@latest` is found in packages list. It does mean that `dlx pkg@latest` will always be slow, but the user should know what they are doing anyway.

There is another problem comes with it: `@latest` perhaps parse as older cached version in my example.
Is possible to parse `@latest` as newer version when install newer version ?

### My opinion:
If `pkg-name@latest` == `pkg-name@0.0.1` because of cache
and pkg-name@0.0.2 has `latest` tag
When the command `pnpm dlx pkg-name@0.0.2` created new cache, `@latest` could be parsed to `@0.0.2`
Otherwise, `@latest` always refers specifically to `@0.0.1`, it's kind of wired.

**Comment by KSXGitHub** [comment]
https://github.com/pnpm/pnpm/pull/8722
